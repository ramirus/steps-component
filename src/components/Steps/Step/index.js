import React from 'react'
import styled, { css } from 'styled-components';

// template props
const tp = {
  activeColor: '#5025D6',
  secondColor: '#DBDBDB',
  iconSize: '40px',
}

const ButtonContainer = styled.div`
    position: relative;

    &:after {
      content: '';
      height: 8px;
      width: 100%;
      background-color: ${props => props.checked ? 'transparent' : tp.secondColor};
      display: block;
      position: absolute;
      top: 16px;
      left: 50%;
      z-index: -1;
      background: linear-gradient(to right, ${props => `${tp.activeColor} 50%, ${tp.secondColor} 50%`});
      background-size: 200% 100%;
      background-position: right bottom;
      transition: background-position 0.2s ease 0.1s;
      
      ${
  props => props.checked ?
    css`background-position: left bottom;
    transition: background-position 0.2s`
    :
    ''
  }
    }
  `

const Container = styled.div`
    flex: 1;
    width: 100%;
    min-width: 100px;
    text-align: center;

    &:last-child {
      ${ButtonContainer}: {
        &:after {
          width: 0;
        }
        &:after {
          width: 0;
        }
      }
    }
  `

const Label = styled.label`
    display: block;
    padding-bottom: 15px;
    font-size: 18px;
    font-weight: bold;
    cursor: ${props => props.active ? 'pointer' : 'default'};
    color: ${props => props.checked ? tp.activeColor : tp.secondColor};
    transition: color 0.1s;

    ${props => props.current ?
    css`color: ${props => tp.activeColor};
      transition: color 0.1s ease 0.2s;`
    :
    ''
  }
  `
Label.displayName = 'Label'

const Button = styled.input`
    width: ${tp.iconSize};
    height: ${tp.iconSize};
    border: 10px solid ${props => props.isChecked ? tp.activeColor : tp.secondColor};
    border-radius: 50%;
    padding: 0;
    margin: 0;
    background-color: #fff;
    cursor: ${props => props.active ? 'pointer' : 'default'};
    transition: transform 0.3s;
    transition: border 0.1s;

    ${props => props.active ?
    `&:hover {
        transform: scale(1.2);
      }`
    : ''
  }
  ${props => props.current ?
    css`border: 10px solid ${props => tp.activeColor};
    transition: border 0.1s ease 0.2s`
    :
    ''
  }
  }
    &:focus {
      outline: none;
      transform: scale(1.2);
    }
  `
Button.displayName = 'Button'

const Step = (props) => {
  const handleClick = (e) => {
    e.preventDefault();
    props.handleClick(props.index);
  }

  const {
    checked,
    active,
    current,
  } = props;

  return (
    <Container>
      <Label
        htmlFor={props.title}
        checked={checked}
        active={active}
        current={current}
      >
        {props.title}
      </Label>
      <ButtonContainer checked={checked}>
        <Button
          id={props.title}
          type="button"
          onClick={handleClick}
          disabled={!props.active}
          isChecked={checked}
          active={active}
          current={current}
        />
      </ButtonContainer>
    </Container>
  )
}

export default Step
