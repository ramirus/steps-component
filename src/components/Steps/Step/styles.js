import styled, { css } from 'styled-components';

// template props
const tp = {
  activeColor: '#5025D6',
  secondColor: '#DBDBDB',
  iconSize: '40px',
}

export const ButtonContainer = styled.div`
    position: relative;

    &:after {
      content: '';
      height: 8px;
      width: 100%;
      background-color: ${props => props.checked ? 'transparent' : tp.secondColor};
      display: block;
      position: absolute;
      top: 16px;
      left: 50%;
      z-index: -1;
      background: linear-gradient(to right, ${props => `${tp.activeColor} 50%, ${tp.secondColor} 50%`});
      background-size: 200% 100%;
      background-position: right bottom;
      transition: background-position 0.2s ease 0.1s;
      
      ${props => props.checked ?
    css`background-position: left bottom;
        transition: background-position 0.2s`
    :
    ''
  }
    }
  `

export const Container = styled.div`
    flex: 1;
    width: 100%;
    min-width: 100px;
    text-align: center;

    &:last-child {
      ${ButtonContainer}: {
        &:after {
          width: 0;
        }
        &:after {
          width: 0;
        }
      }
    }
  `

export const Label = styled.label`
    display: block;
    padding-bottom: 15px;
    font-size: 18px;
    font-weight: bold;
    cursor: ${props => props.active ? 'pointer' : 'default'};
    color: ${props => props.checked ? tp.activeColor : tp.secondColor};
    transition: color 0.1s;

    ${props => props.current ?
    css`color: ${props => tp.activeColor};
        transition: color 0.1s ease 0.2s;`
    :
    ''
  }
  `
Label.displayName = 'Label'

export const Button = styled.input`
    width: ${tp.iconSize};
    height: ${tp.iconSize};
    border: 10px solid ${props => props.isChecked ? tp.activeColor : tp.secondColor};
    border-radius: 50%;
    padding: 0;
    margin: 0;
    background-color: #fff;
    cursor: ${props => props.active ? 'pointer' : 'default'};
    transition: transform 0.3s;
    transition: border 0.1s;

    ${props => props.active ?
    `&:hover {
          transform: scale(1.2);
        }`
    :
    ''
  }
    ${props => props.current ?
    css`border: 10px solid ${props => tp.activeColor};
      transition: border 0.1s ease 0.2s`
    :
    ''
  }
  }
    &:focus {
      outline: none;
      transform: scale(1.2);
    }
  `
Button.displayName = 'Button'
