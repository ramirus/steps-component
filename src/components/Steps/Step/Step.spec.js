import React from 'react'
import { shallow } from 'enzyme'

import Step from './index'

describe('Step component', () => {
  const props = {
    title: '',
    handleClick: () => { },
  }

  describe('Step component initial', () => {
    const step = shallow(<Step {...props} />)

    it('renders properly', () => {
      expect(step).toMatchSnapshot()
    })
  })

  describe('Step component with props', () => {
    const mockClickHandle = jest.fn();

    const nextProps = {
      title: 'Test',
      handleClick: mockClickHandle,
    }
    const step = shallow(<Step {...nextProps} />)

    it('renders label', () => {
      expect(step.find('Label')).toHaveLength(1)
    })

    it('renders icon button', () => {
      expect(step.find('input[type="button"]')).toHaveLength(1)
    })

    describe('when clicking the icon button', () => {
      step.find('input[type="button"]').simulate('click', {
        preventDefault: () => { },
      })

      it('calls the props.clickHandle', () => {
        expect(mockClickHandle).toHaveBeenCalledTimes(1)
      })
    })
  })
})
