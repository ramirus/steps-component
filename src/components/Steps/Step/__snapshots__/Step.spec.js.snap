// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step component Step component initial renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Step
    handleClick={[Function]}
    title=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <ForwardRef
          htmlFor=""
        >
          
        </ForwardRef>,
        <ForwardRef>
          <ForwardRef
            disabled={true}
            id=""
            onClick={[Function]}
            type="button"
          />
        </ForwardRef>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "active": undefined,
          "checked": undefined,
          "children": "",
          "current": undefined,
          "htmlFor": "",
        },
        "ref": null,
        "rendered": "",
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-htpNat",
            "isStatic": false,
            "rules": Array [
              "
    display: block;
    padding-bottom: 15px;
    font-size: 18px;
    font-weight: bold;
    cursor: ",
              [Function],
              ";
    color: ",
              [Function],
              ";
    transition: color 0.1s;

    ",
              [Function],
              "
  ",
            ],
          },
          "displayName": "Label",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-htpNat",
          "target": "label",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "checked": undefined,
          "children": <ForwardRef
            disabled={true}
            id=""
            onClick={[Function]}
            type="button"
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "active": undefined,
            "current": undefined,
            "disabled": true,
            "id": "",
            "isChecked": undefined,
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bxivhb",
              "isStatic": false,
              "rules": Array [
                "
    width: ",
                "40px",
                ";
    height: ",
                "40px",
                ";
    border: 10px solid ",
                [Function],
                ";
    border-radius: 50%;
    padding: 0;
    margin: 0;
    background-color: #fff;
    cursor: ",
                [Function],
                ";
    transition: transform 0.3s;
    transition: border 0.1s;

    ",
                [Function],
                "
  ",
                [Function],
                "
  }
    &:focus {
      outline: none;
      transform: scale(1.2);
    }
  ",
              ],
            },
            "displayName": "Button",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bxivhb",
            "target": "input",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "rules": Array [
              "
    position: relative;

    &:after {
      content: '';
      height: 8px;
      width: 100%;
      background-color: ",
              [Function],
              ";
      display: block;
      position: absolute;
      top: 16px;
      left: 50%;
      z-index: -1;
      background: linear-gradient(to right, ",
              [Function],
              ");
      background-size: 200% 100%;
      background-position: right bottom;
      transition: background-position 0.2s ease 0.1s;
      
      ",
              [Function],
              "
    }
  ",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-bwzfXH",
        "isStatic": true,
        "rules": Array [
          "
    flex: 1;
    width: 100%;
    min-width: 100px;
    text-align: center;

    &:last-child {
      ",
          ".sc-bdVaJa",
          ": {
        &:after {
          width: 0;
        }
        &:after {
          width: 0;
        }
      }
    }
  ",
        ],
      },
      "displayName": "styled.div",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-bwzfXH",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <ForwardRef
            htmlFor=""
          >
            
          </ForwardRef>,
          <ForwardRef>
            <ForwardRef
              disabled={true}
              id=""
              onClick={[Function]}
              type="button"
            />
          </ForwardRef>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "active": undefined,
            "checked": undefined,
            "children": "",
            "current": undefined,
            "htmlFor": "",
          },
          "ref": null,
          "rendered": "",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-htpNat",
              "isStatic": false,
              "rules": Array [
                "
    display: block;
    padding-bottom: 15px;
    font-size: 18px;
    font-weight: bold;
    cursor: ",
                [Function],
                ";
    color: ",
                [Function],
                ";
    transition: color 0.1s;

    ",
                [Function],
                "
  ",
              ],
            },
            "displayName": "Label",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-htpNat",
            "target": "label",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "checked": undefined,
            "children": <ForwardRef
              disabled={true}
              id=""
              onClick={[Function]}
              type="button"
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "active": undefined,
              "current": undefined,
              "disabled": true,
              "id": "",
              "isChecked": undefined,
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": null,
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-bxivhb",
                "isStatic": false,
                "rules": Array [
                  "
    width: ",
                  "40px",
                  ";
    height: ",
                  "40px",
                  ";
    border: 10px solid ",
                  [Function],
                  ";
    border-radius: 50%;
    padding: 0;
    margin: 0;
    background-color: #fff;
    cursor: ",
                  [Function],
                  ";
    transition: transform 0.3s;
    transition: border 0.1s;

    ",
                  [Function],
                  "
  ",
                  [Function],
                  "
  }
    &:focus {
      outline: none;
      transform: scale(1.2);
    }
  ",
                ],
              },
              "displayName": "Button",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-bxivhb",
              "target": "input",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            },
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bdVaJa",
              "isStatic": false,
              "rules": Array [
                "
    position: relative;

    &:after {
      content: '';
      height: 8px;
      width: 100%;
      background-color: ",
                [Function],
                ";
      display: block;
      position: absolute;
      top: 16px;
      left: 50%;
      z-index: -1;
      background: linear-gradient(to right, ",
                [Function],
                ");
      background-size: 200% 100%;
      background-position: right bottom;
      transition: background-position 0.2s ease 0.1s;
      
      ",
                [Function],
                "
    }
  ",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bdVaJa",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bwzfXH",
          "isStatic": true,
          "rules": Array [
            "
    flex: 1;
    width: 100%;
    min-width: 100px;
    text-align: center;

    &:last-child {
      ",
            ".sc-bdVaJa",
            ": {
        &:after {
          width: 0;
        }
        &:after {
          width: 0;
        }
      }
    }
  ",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-bwzfXH",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
